telemetry/telemetry.go

```diff
@@ -0,0 +1,263 @@
+	if err != nil {
+		logger.Fatal("err", err)
+	}
+
+	name := "" // Using default name for provider
+	tracer := otel.Tracer(name)
+	meter := otel.Meter(name)
+
+	return Telemetry{
+		Tracer:                    tracer,
+		Meter:                     meter,
+		TelemetryDataByMerkleRoot: make(map[[32]byte]TraceData),
+		dataMutex:                 &sync.Mutex{},
+		initialCtx:                ctx,
+		shutdownTracerProvider:    shutdownTracerProvider,
+		shutdownMeterProvider:     shutdownMeterProvider,
+	}
+}
+
+// Telemetry functions
+
+// InitNewTrace Init a new trace for the given batchMerkleRoot
+// User must make sure to call FinishTrace()
+func (t *Telemetry) InitNewTrace(batchMerkleRoot [32]byte) {
+	merkleRootString := hex.EncodeToString(batchMerkleRoot[:])
+	ctx, span := t.Tracer.Start(
+		context.Background(),
+		fmt.Sprintf("Response for 0x%s", merkleRootString),
+		trace.WithAttributes(
+			attribute.String("merkle_root", fmt.Sprintf("0x%s", merkleRootString)),
+		),
+	)
+	t.dataMutex.Lock()
+	defer t.dataMutex.Unlock()
+	t.TelemetryDataByMerkleRoot[batchMerkleRoot] = TraceData{
+		Ctx:  ctx,
+		Span: span,
+	}
+}
+
+// FinishTrace finishes the trace for the given merkle root and frees resources
+// In order to wait for all operators responses, even if the quorum is reached, this function has a delayed execution
+func (t *Telemetry) FinishTrace(batchMerkleRoot [32]byte) {
+	go func() {
+		time.Sleep(10 * time.Second)
+		span := t.getSpan(batchMerkleRoot)
+		span.End()
+		t.dataMutex.Lock()
+		defer t.dataMutex.Unlock()
+		delete(t.TelemetryDataByMerkleRoot, batchMerkleRoot)
+	}()
+}
+
+// OperatorResponseTrace
+// User must call to `defer span.End()` to make sure the span is correctly finished
+// For example
+// ```
+//
+//	span := telemetry.OperatorResponseTrace(batchMerkleRoot, operatorId)
+//	defer span.End()
+//
+// ```
+func (t *Telemetry) OperatorResponseTrace(batchMerkleRoot [32]byte, operatorId [32]byte) trace.Span {
+	ctx := t.getCtx(batchMerkleRoot)
+	operatorIdString := hex.EncodeToString(operatorId[:])
+	_, span := t.Tracer.Start(
+		ctx,
+		fmt.Sprintf("Operator ID: 0x%s", operatorIdString),
+		trace.WithAttributes(
+			attribute.String("merkle_root", fmt.Sprintf("0x%s", hex.EncodeToString(batchMerkleRoot[:]))),
+			attribute.String("operator_id", fmt.Sprintf("0x%s", operatorIdString)),
+		),
+	)
+	return span
+}
+
+// QuorumReachedTrace
+// User must call to `defer span.End()` to make sure the span is correctly finished
+// For example
+// ```
+//
+//	span := telemetry.QuorumReachedTrace(batchMerkleRoot)
+//	defer span.End()
+//
+// ```
+func (t *Telemetry) QuorumReachedTrace(batchMerkleRoot [32]byte) trace.Span {
+	ctx := t.getCtx(batchMerkleRoot)
+	_, span := t.Tracer.Start(
+		ctx,
+		"Quorum reached",
+		trace.WithAttributes(
+			attribute.String("merkle_root", fmt.Sprintf("0x%s", hex.EncodeToString(batchMerkleRoot[:]))),
+			attribute.String("status", "ok"),
+		),
+	) // TODO add quorum %
```

Comment:
Should this TODO be resolved in this PR?

Reply:
This will be added in another pr